lib/api/completeMultipartUpload.js

```diff
@@ -524,15 +524,64 @@ function completeMultipartUpload(authInfo, request, log, callback) {
+                keysToDelete, log, err => {
+                    if (err) {
+                        log.error('error deleting MPU metadata during completeMPU', {
+                            method: 'completeMultipartUpload',
+                            mpuBucket: mpuBucket.getName(),
+                            keysToDeleteCount: keysToDelete.length,
+                            error: err,
+                        });
+
+                        // Handle specific error cases according to retry strategy
+                        if (err.is?.DeleteConflict) {
+                            // DeleteConflict should trigger automatic retry
+                            // Convert to InternalError to make it retryable
+                            return next(errors.InternalError, extraPartLocations,
+                                destinationBucket, aggregateETag, generatedVersionId, droppedMPUSize);
+                        }
+
+                        // For NoSuchKey and other errors, return them as-is
+                        // NoSuchKey is non-retryable, InternalError and others are retryable
+                        return next(err, extraPartLocations,
+                            destinationBucket, aggregateETag, generatedVersionId, droppedMPUSize);
+                    }
+
+                    log.debug('successfully deleted MPU metadata during completeMPU', {
+                        method: 'completeMultipartUpload',
+                        mpuBucket: mpuBucket.getName(),
+                        keysToDeleteCount: keysToDelete.length,
+                    });
+                    return next(null, extraPartLocations,
+                        destinationBucket, aggregateETag, generatedVersionId, droppedMPUSize);
+                });
         },
         function batchDeleteExtraParts(extraPartLocations, destinationBucket,
             aggregateETag, generatedVersionId, droppedMPUSize, next) {
             if (extraPartLocations && extraPartLocations.length > 0) {
                 return data.batchDelete(extraPartLocations, request.method, null, log, err => {
                     if (err) {
-                        return next(err);
+                        // Extra part deletion failure should not fail the operation
+                        // The S3 object was created successfully and MPU metadata was cleaned up
+                        // Orphaned extra parts are acceptable since the main operation succeeded
+                        log.warn('failed to delete extra parts, keeping orphan but returning success', {
+                            method: 'completeMultipartUpload',
+                            extraPartLocationsCount: extraPartLocations.length,
+                            error: err,
+                        });
+
+                        // Continue with quota validation but don't fail if it errors either
+                        return validateQuotas(request, destinationBucket, request.accountQuotas,
```

Comment:
Is it doable to make the call to `validateQuotas` an extra step in the waterflow, to avoid duplicating the code? I don't think we really care that the log doesn't specify `after extra part deletion failure` when failing to update quotas.

Reply:
Yes it is, now that we do not return any error, I applied the suggestion :slightly_smiling_face: